File: src/libs/cpp/actions/throw_exception.cpp
#include "throw_exception.hpp"

#include <iostream>

void throw_exception(VirtualMachine &vm) noexcept
{
	auto err = std::get<std::map<std::string, Value>>(vm.stack.back());
	auto &err_stack = std::get<std::vector<Value>>(err["stack"]);

	if (vm.except_stack.empty())
	{
		// If exception is not caught, end the program immediately and output the error.
		auto message = std::get<std::string>(err["message"]);
		auto line = (int)std::get<double>(err["line"]);
		std::cerr << "ERROR: [line " << line << "] " << message << std::endl;
		std::cerr << "Error not caught, program terminated." << std::endl;
		std::exit(1);
	}

	// Otherwise, we are catching exceptions.

	// Unroll program stack
	auto &info = vm.except_stack.back();
	vm.except_stack.pop_back();
	vm.stack.resize(info.stack_size);

	// Unroll call stack
	while (vm.return_indices.size() > info.return_stack_size)
	{
		auto retn = vm.return_indices.back();
		vm.return_indices.pop_back();
		auto line = vm.instructions[retn.index].line_no;
		std::get<std::vector<Value>>(err["line"]).push_back(line);
	}

	vm.instruction_index = info.goto_index - 1;
	err["line"] = err_stack.back();

	vm.stack.push(err);
}

File: src/runtime/actions/throw_exception.lua
return function(vm, line, p1, p2)
	local err = STACK[#STACK]

	if #EXCEPT_STACK == 0 then
		--If exception is not caught, end the program immediately and output the error
		CURRENT_INSTRUCTION = #INSTRUCTIONS + 1

		local msg = ''

		---@diagnostic disable-next-line
		if FILE and #FILE > 0 then
			msg = '["' .. err.file .. '": ' .. err.line .. '] ' .. err.message
		else
			msg = '[line ' .. err.line .. '] ' .. err.message
		end
		output_array({ "error", 'ERROR: ' .. msg .. '\nError not caught, program terminated.' }, 7)

		return
	end

	--Otherwise, we are catching exceptions.

	--Unroll program stack
	local catch = table.remove(EXCEPT_STACK)
	while #STACK > catch.stack do
		table.remove(STACK)
	end

	--Unroll call stack
	while #INSTR_STACK > catch.instr_stack do
		table.remove(INSTR_STACK) --Remove any function parameters
		table.remove(INSTR_STACK) --Remove stack size value
		local instr_id = table.remove(INSTR_STACK)
		table.insert(err.stack, 1, INSTRUCTIONS[instr_id][2])
	end
	CURRENT_INSTRUCTION = catch.instr

	err.line = err.stack[#err.stack]

	vm.push(err)
end
